Refactors and minor improvements related to cluster queries

nrwahl2 opened the pull request:
My stack of commits related to the `crm_mon` disconnect stuff ([T15](https://projects.clusterlabs.org/T15)) has been growing. I realized that a bunch of them have merit on their own independent of the `crm_mon` disconnect work. I'm opening this PR now to get those reviewed and make eventual review of the rest easier.

I'm not super tied to the "Functionize reply validation" commit. It's a little bit clunky because of the need for different type casts. But it does reduce duplication, and that will be even more pronounced if we ever add more daemons to `pcmk_cluster_queries.c`.

nrwahl2 commented on lib/pacemaker/pcmk_cluster_queries.c:
```diff
@@ -208,6 +208,17 @@ designated_controller_event_cb(pcmk_ipc_api_t *controld_api,
 
+/*!
+ * \internal
+ * \brief Process a \p pacemakerd status IPC event
```
A completely unimportant stylistic note: I was torn on whether to use "`pacemakerd`" or "launcher" in doxygen blocks. We usually say "controller" and "scheduler" instead of "`pacemaker-controld` and "`pacemaker-schedulerd`". "Launcher" is the analog for "`pacemakerd`", and we use it in log messages (`pcmk_ipc_name()`).

However, it's used almost nowhere in the code or documentation right now so its meaning is probably less clear than "`pacemakerd`".

kgaillot commented on include/crm/common/ipc_controld.h:
```diff
@@ -37,6 +37,8 @@ enum pcmk_controld_api_reply {
 };
 
+const char *pcmk__controld_api_reply2str(enum pcmk_controld_api_reply reply);
+
```
If these are meant to be internal, they should go in ipc_internal.h (or new ipc_*_internal.h files); if they're meant to be public, single underbar

nrwahl2 replied:
whoops

Definitely private.

kgaillot commented on lib/pacemaker/pcmk_cluster_queries.c:
```diff
@@ -483,10 +491,11 @@ pcmk_designated_controller(xmlNodePtr *xml, unsigned int message_timeout_ms)
+ *       returns \p pcmk_rc_ok if the IPC connection attempt returns
+ *       \p EREMOTEIO. That code indicates that this is a Pacemaker Remote node
+ *       with \p pacemaker-remoted running. The node may be connected to the
+ *       cluster.
```
This same change would need to be made for the public equivalent.

I'm ambivalent about it. It's technically breaking API compatibility since this return code was documented, but that's not a dealbreaker since this API is so new that I don't think it's gotten any external use yet. On the one hand, this is convenient for callers, but on the other hand, it's claiming something it doesn't know, and distinguishing pacemakerd vs pacemaker-remoted might one day be useful in some context.

nrwahl2 replied:
> this return code was documented

Kind of. The specific handling of `EREMOTEIO` was documented only for the internal function. The public doxygen just says "Standard Pacemaker return code" and says `"Get and output \p pacemakerd status"` (which is not quite the same as `pacemakerd-remote` status, granted). It doesn't give any details about when it will return which specific code.

> it's claiming something it doesn't know

It does know that `pacemaker-remoted` is running at least, which is the nearest meaningful analog we have for `pacemakerd` running on a Pacemaker Remote node. The hope of the doxygen note was to avoid promising something that isn't true.

This is a little bit of a hack, just so convenient. It's not a big deal to revert this though.

nrwahl2 replied:
Hmm, how would you feel about adding a `pcmk_pacemakerd_state_remote` enum value? It's not technically a `pacemakerd` state, but y'know... potentially useful indicator

I just remembered another reason why I did this: to allow `pcmk_cluster_queries.c:pacemakerd_event_cb()` to print something (besides "invalid") for the `pacemakerd` state.

kgaillot replied:
Ah I missed that EREMOTEIO isn't documented for the public API, that makes it less a problem. I like the idea of adding a new enum value so we're giving accurate information.

kgaillot commented on lib/pacemaker/pcmk_status.c:
```diff
@@ -287,19 +287,8 @@ pcmk__status(pcmk__output_t *out, cib_t *cib,
 done:
-    if (stonith != NULL) {
-        if (stonith->state != stonith_disconnected) {
-            stonith->cmds->remove_notification(stonith, NULL);
```
It's helpful to remove all notifications *before* disconnecting; it actually unregisters the notification with the fencer, rather than just deleting the client-side notification table.

Quite possibly we should do that in stonith_api_free() (or maybe only stonith_api_delete()). Offhand, I can't think of a reason not to, except maybe to avoid a timeout if the connection is known to be not working properly.

nrwahl2 replied:
Would checking `stonith->state != stonith_disconnected` within free/delete before unregistering be sufficient?

nrwahl2 replied:
And on the fence between free and delete... at this point delete is wholly redundant to free. Could be deprecated unless we have plans for it in the future.

nrwahl2 replied:
> Could be deprecated

Never mind, it has the NULL check :)

kgaillot replied:
Free and delete are redundant but both are public API so we can't do anything about it :-/ Also, one returns void and the other int, which isn't significant but prevents us from at least reducing it to one actual function.

My gut feeling is that free should unregister the notifications before disconnecting, but I haven't traced through all the callers to be sure that's OK.

kgaillot replied:
The null check should moved to free, there's no reason to have that as a point of difference

nrwahl2 replied:
`stonith_api_free()` is static and always called as a method outside of `st_client.c`, so the argument **should** never be null there.

Also, in theory if there were more than one implementation of the free method, we could still want `stonith_api_delete()` as a wrapper that does a null check. We have other free() wrappers that just null-check scattered around IIRC.

nrwahl2 replied:
> My gut feeling is that free should unregister the notifications before disconnecting

I would at least *expect* that to be okay, provided the client is not already disconnected

nrwahl2 commented on lib/pacemaker/pcmk_cluster_queries.c:
```diff
@@ -331,6 +262,57 @@ ipc_connect(data_t *data, enum pcmk_ipc_server server, pcmk_ipc_callback_t cb,
+        end_nsec = qb_util_nano_current_get();
+        elapsed_ms = (end_nsec - start_nsec) / QB_TIME_NS_IN_MSEC;
+        remaining_ms = data->message_timeout_ms - elapsed_ms;
+    }
```
Good point. That shouldn't really happen currently. The poll timeout is always `remaining_ms`, and we keep polling with `remaining_ms` until we either get the reply or run out of time. The poll is the main thing that should take time here, so a timeout should correspond with a polling timeout.

However, I agree it's a good idea to add a final error message outside the loop if `!data->reply_received`. It is possible. Say we start with a 5000 ms timeout and `pcmk_poll_ipc()` returns after 3000 seconds with data in the buffer. Then say it takes another 3000 seconds to dispatch the message(s), and we find out we haven't received the reply.

nrwahl2 commented:
Updated to address most of the review:
* Moved reply2str functions to internal header
* Unregistered stonith notifications in `stonith_api_free()` (seems reasonable since we already disconnect there, and `stonith_api_delete()` is basically a wrapper for null-safety)
* Log an error in `poll_until_reply()` for timeout from the whole while loop, instead of just from `pcmk_poll_ipc()`.

For now I left the `EREMOTEIO` commit. Still need to figure out what to do with that. I'm thinking a `pcmk_pacemakerd_state_remote` might be the right balance between convenience and clarity, as mentioned in another comment.

nrwahl2 commented:
changes seemed simple enough to just push... silly me.

nrwahl2 commented:
* Added missing include to ipc_internal.h (caused compile failure)
* Added new `pcmk_pacemakerd_state_remote` enum value in a new commit
* Set state to `pcmk_pacemakerd_state_remote` instead of `pcmk_pacemakerd_state_running` for `EREMOTEIO`. I'm doing this regardless of whether `eremoteio_ok` is true, since AFAIK that return code does indicate that `pacemaker-remoted` is running.
* Log a "pacemakerd-health" message outside of the callback if we're on a remote node.

nrwahl2 commented:
Fixed bug introduced in functionized `validate_event_reply()`

nrwahl2 commented:
retest this please

kgaillot commented:
Awesome, thanks :)
